File: mobile/app/src/main/java/cz/kvalitacena/ui/common/CategoryTree.kt
package cz.kvalitacena.ui.common

import cz.kvalitacena.network.Category
import java.text.Collator
import java.util.Locale

/**
 * Zrcadlo webového `shared/category-tree.ts` — viz tam pro celkové odůvodnění. `Query.categories`
 * vrací plochý seznam řazený podle `path` (abecedně podle slugu), hierarchii nese `path` samo
 * (řetězec slugů oddělený „/") a pořadí SOUROZENCŮ v jedné větvi `sortOrder` — appka si strom
 * skládá sama, ne server. Čistá logika mimo Compose (JUnit, ne instrumentace), stejný vzor jako
 * `PublicationStatusText.kt`/`PriceRowValidation.kt`. Používá `ui/product/ProductFormScreen.kt`
 * přes generický `SearchableDropdown` (dokud uživatel nic nenapsal, nabídne strom v pořadí
 * stromu s odsazením podle hloubky; při psaní přepne na plochý filtr s breadcrumb popiskem).
 */
data class CategoryTreeItem(val category: Category, val depth: Int)

private fun parentPath(path: String): String? {
  val idx = path.lastIndexOf('/')
  return if (idx == -1) null else path.substring(0, idx)
}

private fun siblingComparator(locale: Locale): Comparator<Category> {
  val collator = Collator.getInstance(locale)
  return compareBy<Category> { it.sortOrder }.thenComparator { a, b -> collator.compare(a.name, b.name) }
}

/**
 * Poskládá plochý seznam do pořadí stromu (rodič, pak potomci seřazení podle (sortOrder, name),
 * rekurzivně) s dopočítanou hloubkou pro odsazení. Vnitřní uzly (mají potomky) jsou v seznamu
 * stejně vybíratelné jako listy — kdo pro své zboží nenajde podkategorii, zvolí nadřazenou místo
 * aby zakládání vzdal (docs/nasazeni.md, kap. 4, "Než pozvat první lidi").
 */
fun buildCategoryTreeOrder(
  categories: List<Category>,
  locale: Locale = Locale.getDefault(),
): List<CategoryTreeItem> {
  val comparator = siblingComparator(locale)
  val byPath = categories.associateBy { it.path }
  val childrenByParentPath = mutableMapOf<String, MutableList<Category>>()
  val roots = mutableListOf<Category>()
  for (category in categories) {
    val parent = parentPath(category.path)
    if (parent != null && byPath.containsKey(parent)) {
      childrenByParentPath.getOrPut(parent) { mutableListOf() }.add(category)
    } else {
      roots.add(category)
    }
  }

  val result = mutableListOf<CategoryTreeItem>()
  fun visit(category: Category, depth: Int) {
    result.add(CategoryTreeItem(category, depth))
    childrenByParentPath[category.path]?.sortedWith(comparator)?.forEach { visit(it, depth + 1) }
  }
  roots.sortedWith(comparator).forEach { visit(it, 0) }
  return result
}

/**
 * Popisek pro řádek při hledání ("Potraviny › Mléčné výrobky › Máslo") — samotný list ("Máslo")
 * by bez zbytku větve nebylo poznat, pod čím v číselníku leží.
 */
fun categoryBreadcrumb(category: Category, categories: List<Category>): String {
  val byPath = categories.associateBy { it.path }
  val chain = mutableListOf(category.name)
  var current: Category? = category
  while (true) {
    val parent = parentPath(current!!.path)?.let { byPath[it] } ?: break
    chain.add(0, parent.name)
    current = parent
  }
  return chain.joinToString(" › ")
}

/** Filtr pro režim hledání v `SearchableDropdown` — case-insensitive podle jména kategorie. */
fun filterCategories(query: String, categories: List<Category>): List<Category> {
  if (query.isBlank()) return emptyList()
  val needle = query.trim()
  return categories.filter { it.name.contains(needle, ignoreCase = true) }
}

private const val TREE_INDENT_PER_LEVEL = "    "

/** Řádek nabídnutý ve `SearchableDropdown<CategoryChoice>` — [label] nese buď odsazení podle
 *  hloubky (dokud uživatel nic nenapsal), nebo breadcrumb celé větve (při hledání). */
data class CategoryChoice(val category: Category, val label: String)

/**
 * Sestaví nabídku pro `SearchableDropdown` (ui/product/ProductFormScreen.kt): prázdný dotaz
 * ukáže celý strom v pořadí sortOrder s odsazením podle hloubky, jinak plochý filtr podle
 * jména s breadcrumb popiskem. `locale` volající bere z Compose kontextu
 * (`LocalConfiguration.current.locales[0]`, ne `Locale.getDefault()` — stejný důvod jako
 * RelativeDate.kt/Money.kt), aby řazení sourozenců reagovalo na přepnutí jazyka za běhu.
 */
fun categoryChoicesFor(
  query: String,
  categories: List<Category>,
  locale: Locale = Locale.getDefault(),
): List<CategoryChoice> =
  if (query.isBlank()) {
    buildCategoryTreeOrder(categories, locale).map {
      CategoryChoice(it.category, TREE_INDENT_PER_LEVEL.repeat(it.depth) + it.category.name)
    }
  } else {
    filterCategories(query, categories).map { CategoryChoice(it, categoryBreadcrumb(it, categories)) }
  }

File: frontend/src/app/shared/category-tree.ts
import type { NzTreeNodeOptions } from 'ng-zorro-antd/core/tree';

/**
 * Položka číselníku kategorií tak, jak ji vrací dotaz `Categories`
 * (`services/product-service.ts`) — plochý seznam, hierarchii nese `path` (řetězec slugů
 * oddělený „/"), pořadí sourozenců v jedné větvi `sortOrder` (backend `schema.graphqls`,
 * `type Category`).
 */
export interface CategoryOption {
  id: string;
  name: string;
  path: string;
  sortOrder: number;
}

function parentPath(path: string): string | null {
  const idx = path.lastIndexOf('/');
  return idx === -1 ? null : path.slice(0, idx);
}

/**
 * Poskládá plochý seznam kategorií do stromu pro `nz-tree-select` (`features/product-form`).
 * `Query.categories` vrací pořadí podle `path` (abecedně podle slugu), NE podle kurátorského
 * `sortOrder` — appka proto sourozence přeskládá sama, přes `Intl.Collator` podle aktuálního
 * jazyka appky (zvládne `č`/`ř`/`ł` líp než jednotná SQL collation napříč lokalizacemi).
 *
 * Vnitřní uzly (mají potomky) jdou vybrat stejně jako listy — kdo pro své zboží nenajde
 * podkategorii, zvolí nadřazenou místo aby zakládání vzdal (docs/nasazeni.md, kap. 4,
 * "Než pozvat první lidi").
 */
export function buildCategoryTree(
  categories: readonly CategoryOption[],
  locale = 'cs',
): NzTreeNodeOptions[] {
  const collator = new Intl.Collator(locale);
  const byPath = new Map<string, CategoryOption>();
  categories.forEach((category) => byPath.set(category.path, category));

  const childrenByParentPath = new Map<string, CategoryOption[]>();
  const roots: CategoryOption[] = [];
  for (const category of categories) {
    const parent = parentPath(category.path);
    if (parent !== null && byPath.has(parent)) {
      const siblings = childrenByParentPath.get(parent) ?? [];
      siblings.push(category);
      childrenByParentPath.set(parent, siblings);
    } else {
      roots.push(category);
    }
  }

  const sortSiblings = (items: CategoryOption[]): CategoryOption[] =>
    [...items].sort((a, b) => a.sortOrder - b.sortOrder || collator.compare(a.name, b.name));

  const toNode = (category: CategoryOption): NzTreeNodeOptions => {
    const children = sortSiblings(childrenByParentPath.get(category.path) ?? []).map(toNode);
    return {
      key: category.id,
      title: category.name,
      isLeaf: children.length === 0,
      children: children.length > 0 ? children : undefined,
    };
  };

  return sortSiblings(roots).map(toNode);
}

/**
 * Popisek vybrané kategorie pro uzavřené pole výběru ("Potraviny › Mléčné výrobky › Máslo") —
 * samotný list ("Máslo") by bez zbytku větve nebylo poznat, pod čím v číselníku leží.
 */
export function categoryBreadcrumb(
  categoryId: string,
  categories: readonly CategoryOption[],
): string {
  const byId = new Map(categories.map((category) => [category.id, category]));
  const byPath = new Map(categories.map((category) => [category.path, category]));
  const category = byId.get(categoryId);
  if (!category) return '';

  const chain: string[] = [];
  let current: CategoryOption | undefined = category;
  while (current) {
    chain.unshift(current.name);
    const parent = parentPath(current.path);
    current = parent !== null ? byPath.get(parent) : undefined;
  }
  return chain.join(' › ');
}
